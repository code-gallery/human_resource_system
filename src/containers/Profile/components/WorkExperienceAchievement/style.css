.WorkExperienceAchievement .form-group .remove-row-btn {
  margin-top: 16px;
}
