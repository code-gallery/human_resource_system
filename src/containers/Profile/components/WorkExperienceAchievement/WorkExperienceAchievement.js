import React, { Component } from 'react'
import PropTypes from 'prop-types'
import _isNil from 'lodash/isNil'
import { updateObjectInArray }
  from 'containers/Profile/components/EducationList/components/EducationItemEdit/components/SchoolForm'
import './style.css'

class WorkExperienceAchievement extends Component {
  updateRow = (object, index) => {
    this.props.onFieldChange('achievements',
      updateObjectInArray(
        this.props.achievements,
        { item: object, index }
      )
    )
  }

  removeRow = (index) => {
    this.props.onFieldChange('achievements', this.props.achievements.filter((_, i) => i !== index))
  }

  addRow = () => {
    const { achievements } = this.props
    const emptyRow = { achievement: '' }
    if (achievements && achievements.length) {
      this.props.onFieldChange('achievements', [ ...this.props.achievements, emptyRow ])
    } else {
      this.props.onFieldChange('achievements', [ emptyRow ])
    }
  }

  renderAchievements() {
    const { achievements } = this.props

    if (achievements && achievements.length) {
      return achievements.map((item, index) => {
        return (
          <div className="row" key={index}>
            <div className="col-md-8 col-xs-12">
              <div className="form-group">
                <input
                  type="text"
                  value={item.achievement}
                  onChange={e => this.updateRow({ achievement: e.target.value }, index)}
                />
              </div>
            </div>
            <div className="col-xs-12 col-md-1">
              <div className="form-group">
                <button className="remove-row-btn" onClick={() => {
                  this.removeRow(index)
                }}>&nbsp;</button>
              </div>
            </div>
          </div>
        )
      })
    }

    return null
  }

  render() {
    const { achievements } = this.props
    const showAdd = (_isNil(achievements) || (achievements && achievements.length) < 5)

    return (
      <div className="WorkExperienceAchievement">
        <div className="form-group">
          <p className="help-text">You can add 5 achievements</p>
        </div>

        {this.renderAchievements()}
        {showAdd &&
          <div className="form-group">
            <button className="add-row-btn" onClick={this.addRow}>Add achievement</button>
          </div>
        }
      </div>
    )
  }
}

WorkExperienceAchievement.propTypes = {
  onFieldChange: PropTypes.func,
  achievements: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object
  ])
}

export default WorkExperienceAchievement
