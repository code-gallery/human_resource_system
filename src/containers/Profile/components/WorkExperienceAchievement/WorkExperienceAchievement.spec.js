import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import WorkExperienceAchievement from './WorkExperienceAchievement'

describe('containers/Profile/components/WorkExperienceAchievement', () => {
  const commonProps = {
    onFieldChange: jest.fn()
  }

  describe('when achievements is null', () => {
    it('renders correctly', () => {
      const props = {
        ...commonProps,
        achievements: null
      }
      const tree = renderer.create(
        <WorkExperienceAchievement {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when achievements is not empty', () => {
    it('renders correctly', () => {
      const props = {
        ...commonProps,
        achievements: [
          { achievement: 'hello world' },
          { achievement: 'lorem ipsum' }
        ]
      }
      const tree = renderer.create(
        <WorkExperienceAchievement {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('<WorkExperienceAchievement />', () => {
    let wrapper
    let props
    beforeEach(() => {
      props = {
        ...commonProps,
        achievements: [
          { achievement: 'hello world' },
          { achievement: 'lorem ipsum' }
        ]
      }

      wrapper = shallow(
        <WorkExperienceAchievement {...props} />
      )
    })

    describe('addRow', () => {
      describe('when achievements is not empty', () => {
        it('calls props.onFieldChange correctly', () => {
          wrapper.instance().addRow()
          expect(props.onFieldChange).toHaveBeenCalledWith('achievements', [
            ...props.achievements,
            { achievement: '' }
          ])
        })
      })

      describe('when achievements is empty', () => {
        it('calls props.onFieldChange correctly', () => {
          wrapper.setProps({ achievements: [] })
          wrapper.instance().addRow()
          expect(props.onFieldChange).toHaveBeenCalledWith('achievements', [
            { achievement: '' }
          ])
        })
      })
    })

    describe('removeRow', () => {
      it('calls props.onFieldChange correctly', () => {
        wrapper.instance().removeRow(1)
        expect(props.onFieldChange).toHaveBeenCalledWith('achievements', [
          props.achievements[0]
        ])
      })
    })

    describe('updateRow', () => {
      it('calls props.onFieldChange correctly', () => {
        wrapper.instance().updateRow({ achievement: 'lorem ipsum' }, 1)
        expect(props.onFieldChange).toHaveBeenCalledWith('achievements', [
          props.achievements[0],
          { achievement: 'lorem ipsum' }
        ])
      })
    })
  })
})
