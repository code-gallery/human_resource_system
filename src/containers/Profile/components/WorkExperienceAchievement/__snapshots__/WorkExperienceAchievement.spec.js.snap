// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Profile/components/WorkExperienceAchievement when achievements is not empty renders correctly 1`] = `
<div
  className="WorkExperienceAchievement"
>
  <div
    className="form-group"
  >
    <p
      className="help-text"
    >
      You can add 5 achievements
    </p>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-8 col-xs-12"
    >
      <div
        className="form-group"
      >
        <input
          onChange={[Function]}
          type="text"
          value="hello world"
        />
      </div>
    </div>
    <div
      className="col-xs-12 col-md-1"
    >
      <div
        className="form-group"
      >
        <button
          className="remove-row-btn"
          onClick={[Function]}
        >
           
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-8 col-xs-12"
    >
      <div
        className="form-group"
      >
        <input
          onChange={[Function]}
          type="text"
          value="lorem ipsum"
        />
      </div>
    </div>
    <div
      className="col-xs-12 col-md-1"
    >
      <div
        className="form-group"
      >
        <button
          className="remove-row-btn"
          onClick={[Function]}
        >
           
        </button>
      </div>
    </div>
  </div>
  <div
    className="form-group"
  >
    <button
      className="add-row-btn"
      onClick={[Function]}
    >
      Add achievement
    </button>
  </div>
</div>
`;

exports[`containers/Profile/components/WorkExperienceAchievement when achievements is null renders correctly 1`] = `
<div
  className="WorkExperienceAchievement"
>
  <div
    className="form-group"
  >
    <p
      className="help-text"
    >
      You can add 5 achievements
    </p>
  </div>
  <div
    className="form-group"
  >
    <button
      className="add-row-btn"
      onClick={[Function]}
    >
      Add achievement
    </button>
  </div>
</div>
`;
